dagql/dagui/dot.go

```diff
@@ -0,0 +1,329 @@
+	if !showInternal && vtx.span.Internal {
+		return
+	}
+
+	// hide noisy "id" unless internal requested (even though id is not technically marked as internal)
+	if !showInternal && vtx.call.Field == "id" {
+		return
+	}
+
+	// show it and check its children
+	vtx.show = true
+	for childVtxDgst := range vtx.children {
+		childVtx, ok := dag.vtxByCallDgst[childVtxDgst]
+		if !ok {
+			panic(fmt.Sprintf("child vtx %q not found", childVtxDgst))
+		}
+		dag.setShow(childVtx, visited, showInternal)
+	}
+}
+
+func (dag *dotDag) writeTo(out io.Writer) {
+	fmt.Fprintln(out, "digraph {")
+	defer fmt.Fprintln(out, "}")
+
+	for vtxDgst, vtx := range dag.vtxByCallDgst {
+		if !vtx.show {
+			continue
+		}
+
+		buf := new(bytes.Buffer)
+		fmt.Fprintf(buf, "%s", vtx.call.Field)
+		for ai, arg := range vtx.call.Args {
+			if ai == 0 {
+				fmt.Fprintf(buf, "(")
+			} else {
+				fmt.Fprintf(buf, ", ")
+			}
+			fmt.Fprintf(buf, "%s: %s", arg.Name, displayLit(arg.Value))
+			if ai == len(vtx.call.Args)-1 {
+				fmt.Fprintf(buf, ")")
+			}
+		}
+		if vtx.call.Nth != 0 {
+			fmt.Fprintf(buf, "#%d", vtx.call.Nth)
+		}
+		label := buf.String()
+
+		duration := vtx.span.ActiveDuration(time.Now())
+		label += fmt.Sprintf("\n%s", duration)
+
+		thicc := false
+		if s := duration.Seconds(); s > 1.0 {
+			thicc = true
+		}
+		border := 1.0
+		color := "black"
+		if thicc {
+			border = 10.0
+			color = "red"
+		}
+
+		fmt.Fprintf(out, "  %q [label=%q shape=ellipse penwidth=%f color=%s];\n", vtxDgst, label, border, color)
+
+		for childVtxDgst, edge := range vtx.children {
+			childVtx, ok := dag.vtxByCallDgst[childVtxDgst]
+			if !ok {
+				panic(fmt.Sprintf("child vtx %q not found", childVtxDgst))
+			}
+			if !childVtx.show {
+				continue
+			}
+
+			switch edge.kind {
+			case edgeKindReceiver:
+				fmt.Fprintf(out, "  %q -> %q [color=black];\n", vtx.dgst, childVtx.dgst)
+			case edgeKindArg:
+				fmt.Fprintf(out, "  %q -> %q [color=blue label=%q];\n",
+					vtx.dgst,
+					childVtx.dgst,
+					edge.argName,
+				)
+			case edgeKindSpan:
+				fmt.Fprintf(out, "  %q -> %q [color=green];\n", vtx.dgst, childVtx.dgst)
+			}
+		}
+	}
+}
+
+type edgeKind int
+
+const (
+	edgeKindUnset edgeKind = iota
+	edgeKindReceiver
+	edgeKindArg
+	edgeKindSpan
+)
+
+func findParentSpanWithCall(span *Span) *Span {
+	if span.ParentSpan == nil {
+		return nil
+	}
+	if span.ParentSpan.Call != nil {
+		return span.ParentSpan
+	}
+	return findParentSpanWithCall(span.ParentSpan)
+}
+
+func displayLit(lit *callpbv1.Literal) string {
+	switch val := lit.GetValue().(type) {
+	case *callpbv1.Literal_Bool:
+		return fmt.Sprintf("%v", val.Bool)
+	case *callpbv1.Literal_Int:
+		return fmt.Sprintf("%d", val.Int)
+	case *callpbv1.Literal_Float:
+		return fmt.Sprintf("%f", val.Float)
+	case *callpbv1.Literal_String_:
+		if len(val.String_) > 256 {
+			return "ETOOBIG"
+		}
+		return fmt.Sprintf("%q", val.String_)
+	case *callpbv1.Literal_CallDigest:
+		return "<input>"
```

Comment:
You could almost use the existing `Display() string` code from `dagql/call.Literal` but I guess you might want to avoid the value getting too big with nesting, hence `<input>`. Not a big deal to just roll your own here anyway, since the number of options shouldn't really ever change.

Reply:
Yeah, `<input>` also makes sense since there's an edge drawn for arg inputs that are from a different call, so no need to repeat. Agree copy-pasta is probably better here; also keeps this from having to touch any code that impacts user-facing stuff (would be sad to have a bug related to changes meant only for internal debug tooling)